### PR TITLE
Validate total service DNS hostname+domain length

### LINKED ISSUE #825: Service instance cannot start with too long instance name
Noticed this with a long project name running `kontena app deploy`: service with longer than 48 character name fails to start, so the instance name must be 50 characters or less. Creation and deployment runs fine, but the instance cannot start. I tested this also locally with Docker for Mac (client and server versions `1.12.0-rc2` and api versions `1.24`) and was able to run a container with arbitrary long container name. 

```
$ kontena version
cli: 0.14.3
master: 0.14.3
```

This fails:

```
$ kontena service create --grid dev 0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxxx49 kontena/lb:latest
$ kontena service deploy 0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxxx49
$ kontena service show 0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxxx49
dev/0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxxx49:
  status: running
  image: kontena/lb:latest
  stateful: no
  scaling: 1
  strategy: ha
  deploy_opts:
    min_health: 0.8
  dns: 0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxxx49.dev.kontena.local
  net: bridge
  instances:
    0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxxx49-1:
...
      dns: 0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxxx49-1.dev.kontena.local
...
      status: stopped
      reason: [8] System error: invalid argument
```

This works:

```
$ kontena service create --grid dev 0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxx48 kontena/lb:latest
$ kontena service deploy 0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxx48
$ kontena service show 0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxx48
dev/0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxx48:
  status: running
  image: kontena/lb:latest
  stateful: no
  scaling: 1
  strategy: ha
  deploy_opts:
    min_health: 0.8
  dns: 0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxx48.dev.kontena.local
  net: bridge
  instances:
    0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxx48-1:
...
      dns: 0xxxxxxxxx10xxxxxxxx20xxxxxxxx30xxxxxxxx40xxxx48-1.dev.kontena.local
...
      status: running
```

### PR DESCRIPTION
**Pull request by SpComb**
Docker imposes a limit of 64 chars on the container FQDN (combined `hostname.domain`), because docker/runc will call `sethostname` with the full FQDN. This results in a rather useless Docker API error if the FQDN is too long.

Fixes #825 by failing the service create with a validation error, instead of failing on the service container deploy.

The logic is slightly more complex because it also needs to take into account the number of digits required to represent the container instance number.

## Example

```yaml
stack: test/test-asdfasdfasdfasdf
services:
  xxxxxxxx10xx12:
    image: redis

  xxxxxxxx10xxx13:
    image: redis
    depends_on:
      - xxxxxxxx10xx12
```

#### `kontena stack install test/test-longname.yaml`
```
 [done] Creating stack test-asdfasdfasdfasdf      
 [done] Triggering deployment of stack test-asdfasdfasdfasdf     
 [done] Waiting for deployment to start     
 [done] Deploying service xxxxxxxx10xx12     
 [fail] Deploying service xxxxxxxx10xxx13     
Deployment of service development/test-asdfasdfasdfasdf/xxxxxxxx10xxx13 failed:
  - halting deploy of development/test-asdfasdfasdfasdf/xxxxxxxx10xxx13, one or more instances failed
  - GridServiceInstanceDeployer::ServiceError: Docker::Error::ServerError: invalid header field value "oci runtime error: container_linux.go:247: starting container process caused \"process_linux.go:359: container init caused \\\"invalid argument\\\"\"\n" (on node core-01)
```

#### `kontena stack install test/test-longname.yaml`
```
 [fail] Creating stack test-asdfasdfasdfasdf      
 [error] 422 : Unprocessable Entity:
	services:
	  xxxxxxxx10xxx13:
	    name: 'Total grid service name length 65 is over limit (64): xxxxxxxx10xxx13-1.test-asdfasdfasdfasdf.development.kontena.local'
```

### CODE REVIEW

File: docs/references/kontena-yml.md [hunk 1]
```diff
@@ -52,17 +52,28 @@ services:
...
 services:
   api:
     image: registry/api:latest
   db:
     image: mariadb:latest
 ```
 
+The service name must start with a letter, number or underscore, and may also contain dashes.
+
+There is no limitation on the length of a service name per se, but the resulting service container FQDN (`service-1.stack-name.grid-name.kontena.local`) must be 64 bytes or less. The effective service name length maximum thus also depends on the grid and stack names, so they should be kept reasonably short.
```
Comment (jnummelin) [inline docs/references/kontena-yml.md]: Maybe add note that it's really docker/runc restriction

Comment (SpComb) [inline docs/references/kontena-yml.md]: There isn't really any issue to link to for it, though... probably need to create one first?

File: docs/using-kontena/grids.md [hunk 2]
```diff
@@ -11,10 +11,17 @@ The [Grid](../core-concepts/architecture.md#the-grid) is a top-level object in K
...
 
-## Grid Options
+## Configuration
 
 There are a number of per-grid options. Some of these can be changed using `kontena grid update`, but others must be determined ahead of time when using `kontena grid create`, and cannot be changed later.
 
+
+### Grid name
+
+The grid name will be used as part of the `grid-name.kontena.local` DNS domain used for service lookups. The grid name may consist of letters (`a-z`), digits (`0-9`), underscores (`_`) and dashes (`-`).
```
Comment (SpComb) [inline docs/using-kontena/grids.md]: BTW I wonder how well those `_` would work... or any `-` at the start?

**Comment by SpComb** [comment]
Installing a stack with two services, one under the name limit, the other over, results in an installed stack with missing services... the `Stacks::Create#validate` -> `validate_services` calls `GridServices::Create#validate` with the service definition, but *without* the `stack`, so the service name validation doesn't take the length of the stack name into account, and then it fails during the `create_services` :(

Not sure if worth fixing...

```
$ bin/kontena stack install test/test-longname.yaml 
 [fail] Creating stack test-asdfasdfasdfasdf      
 [error] 422 : Unprocessable Entity:
	services:
	  xxxxxxxx10xxx13:
	    name: 'Total grid service name length 65 is over limit (64): xxxxxxxx10xxx13-1.test-asdfasdfasdfasdf.development.kontena.local'
$ bin/kontena stack show test-asdfasdfasdfasdf
test-asdfasdfasdfasdf:
  created: 2017-09-19T08:47:24.964Z
  updated: 2017-09-19T08:47:24.966Z
  state: initialized
  stack: test/test-asdfasdfasdfasdf
  version: 0.0.1
  revision: 1
  expose: -
  parent: -
  services:
    xxxxxxxx10xx12:
      created: 2017-09-19T08:47:24.976Z
      updated: 2017-09-19T08:47:24.976Z
      image: redis:latest
      revision: 
      state: initialized
      stateful: no
      scaling: 1
      strategy: ha
      read_only: no
      deploy_opts:
        min_health: 
      dns: xxxxxxxx10xx12.test-asdfasdfasdfasdf.development.kontena.local
```

**Pull request merged**